mlflow/utils/gorilla.py

```diff
@@ -353,35 +347,36 @@ def revert(patch):
     with modifictions for autologging disablement purposes.
     """
     # If an curr_active_patch has not been set on destination class for the current patch,
-    # then there is no reverting to do.
+    # then the patch has not been applied and we do not need to revert anything.
     curr_active_patch = _ACTIVE_PATCH % (patch.name,)
-    if not hasattr(patch.destination, curr_active_patch):
+    if curr_active_patch not in patch.destination.__dict__:
+        # already reverted.
         return
 
-    try:
-        original = get_original_attribute(patch.destination, patch.name)
-    except AttributeError:
+    original_name = _ORIGINAL_NAME % (patch.name,)
+
+    # check whether original_name is in destination. We cannot use hasattr because it will
+    # try to get attribute from parent classes if attribute not found in destination class.
+    if original_name not in patch.destination.__dict__:
         raise RuntimeError(
             "Cannot revert the attribute named '%s' since the setting "
             "'store_hit' was not set to True when applying the patch."
             % (patch.destination.__name__,)
         )
 
-    original_name = _ORIGINAL_NAME % (patch.name,)
-    if not getattr(patch.destination, original_name, None):
-        return
+    # during reverting patch, we need restore the raw attribute to the patch point
+    # so get original attribute bypassing descriptor protocal
+    original = object.__getattribute__(patch.destination, original_name)
 
-    setattr(patch.destination, patch.name, original)
+    if patch.is_inplace_patch:
```

Comment:
If the patch is an inplace patch, then won't the attribute with name `original_name` always be absent from the object referred to by the patch?

I.e. if `is_inplace_patch == True`, then `original_name not in patch.destination.__dict__`, right?

If that's true, then we should modify the condition on line 360 to only raise a runtime error if `original_name not in patch.destination.__dict__ and not is_inplace_patch`. Can we also add a test case for this?

Reply:
No. If patched, then always `original_name in patch.destination.__dict__`.  "is_inplace_patch" only means replace the existing method with the patched method, but it will still store the original method to the "original_name" attribute.

e.g.
```
class A:
    def predict(self):
        pass
```

We patch `A.predict`, then it is an in-place patch, and after patch, `A.predict` point to the patched_predict method, and `A._gorilla_original_predict` point to the original `A.predict`

Reply:
Maybe "is_inplace_patch" name is bad, use the name "is_existing_method_replaced_by_patch" looks better ?

Reply:
Are you sure? `original_name` should only be set when there's a hit and `store_hit = True` is passed to `apply()`. https://github.com/mlflow/mlflow/blob/66b4116b6afb66561b81554a1303ef9f11b1adc1/mlflow/utils/gorilla.py#L324-L326